src/coreclr/vm/codeman.cpp

```diff
@@ -4480,120 +4489,49 @@ RangeSection* ExecutionManager::GetRangeSection(TADDR addr)
         SUPPORTS_DAC;
     } CONTRACTL_END;
 
-    RangeSection * pHead = m_CodeRangeList;
-
-    if (pHead == NULL)
+    if (m_RangeSectionHandleArray == nullptr)
     {
         return NULL;
     }
 
-    RangeSection *pCurr = pHead;
-    RangeSection *pLast = NULL;
-
 #ifndef DACCESS_COMPILE
-    RangeSection *pLastUsedRS = (pCurr != NULL) ? pCurr->pLastUsed : NULL;
-
-    if (pLastUsedRS != NULL)
+    int LastUsedRSIndex = (int)m_LastUsedRSIndex;
+    if (LastUsedRSIndex != -1)
     {
-        // positive case
-        if ((addr >= pLastUsedRS->LowAddress) &&
-            (addr <  pLastUsedRS->HighAddress)   )
-        {
-            return pLastUsedRS;
-        }
+        TADDR LowAddress = m_RangeSectionHandleArray[LastUsedRSIndex].LowAddress;
+        TADDR HighAddress = m_RangeSectionHandleArray[LastUsedRSIndex].HighAddress;
 
-        RangeSection * pNextAfterLastUsedRS = pLastUsedRS->pnext;
+        //positive case
+        if ((addr >= LowAddress) && (addr < HighAddress))
+        return m_RangeSectionHandleArray[LastUsedRSIndex].pRS;
 
-        // negative case
-        if ((addr <  pLastUsedRS->LowAddress) &&
-            (pNextAfterLastUsedRS == NULL || addr >= pNextAfterLastUsedRS->HighAddress))
-        {
+        //negative case
+        if ((addr < LowAddress) && ((LastUsedRSIndex == 0)
```

Comment:
You can add the same check for the right side of LastUsedRSIndex, i.e. when addr gets between LastUsedRSIndex and LastUsedRSIndex+1